combining 3 ingest docs (#10177)
opensearch-project/documentation-website: _ingest-pipelines/complex-conditionals.md

## Multiple condition checks

You can combine logical operators like `&&` (and), `||` (or), and `!` (not) to construct more complex conditions. The following pipeline tags documents as `spam` and drops them if they contain an `error_code` higher than `1000`:

```json
PUT _ingest/pipeline/spammy_error_handler
{
  "processors": [
    {
      "set": {
        "field": "tags",
        "value": ["spam"],
        "if": "ctx.message != null && ctx.message.contains('OutOfMemoryError')"
      }
    },
    {
      "drop": {
        "if": "ctx.tags != null && ctx.tags.contains('spam') && ctx.error_code != null && ctx.error_code > 1000"
      }
    }
  ]
}
```
{% include copy-curl.html %}

You can test the pipeline using the following `_simulate` request:

```json
POST _ingest/pipeline/spammy_error_handler/_simulate
{
  "docs": [
    { "_source": { "message": "OutOfMemoryError occurred", "error_code": 1200 } },
    { "_source": { "message": "OutOfMemoryError occurred", "error_code": 800 } },
    { "_source": { "message": "All good", "error_code": 200 } }
  ]
}
```
{% include copy-curl.html %}

> The first document is dropped because it contains an `OutOfMemoryError` string and the `error_code` higher than `1000`:

Reviewer: ```suggestion
The first document is dropped because it contains an `OutOfMemoryError` string and an `error_code` higher than `1000`:
```

```json
PUT _ingest/pipeline/flag_suspicious_ips
{
  "processors": [
    {
      "set": {
        "field": "alert",
        "value": "suspicious_ip",
        "if": "ctx.ip != null && ctx.ip =~ /^192\.168\.\d+\.\d+$/"
      }
    }
  ]
}
```
{% include copy-curl.html %}

Test the pipeline using the following `_simulate` request:

```json
POST _ingest/pipeline/flag_suspicious_ips/_simulate
{
  "docs": [
    { "_source": { "ip": "192.168.0.1" } },
    { "_source": { "ip": "10.0.0.1" } }
  ]
}
```
{% include copy-curl.html %}

The first document has an `alert` field added:

```json
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "alert": "suspicious_ip",
          "ip": "192.168.0.1"
        },
        "_ingest": {
          "timestamp": "2025-04-23T10:32:45.367916428Z"
        }
      }
    },
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "ip": "10.0.0.1"
        },
        "_ingest": {
          "timestamp": "2025-04-23T10:32:45.36793772Z"
        }
      }
    }
  ]
}
```

## Combining fields and null checks

> The following pipeline adds a `priority` field set to `high` if `level` is `critical` and `timestamp` is provided. The script also ensures that all fields are present and meet specific conditions before proceeding

Reviewer: ```suggestion
The following pipeline adds a `priority` field set to `high` if `level` is `critical` and `timestamp` is provided. The script also ensures that all fields are present and meet specific conditions before proceeding:
```

- Adds a `severity` field set to `major` if the value in the `status` field is greater than or equal to `500`.
- Drops the document if the `env` field is set to `test` and the `message` field contains `debug`.

```json
PUT _ingest/pipeline/advanced_log_pipeline
{
  "processors": [
    {
      "set": {
        "field": "env",
        "value": "production",
        "if": "!ctx.containsKey('env')"
      }
    },
    {
      "set": {
        "field": "severity",
        "value": "major",
        "if": "ctx.status != null && ctx.status >= 500"
      }
    },
    {
      "drop": {
        "if": "ctx.env == 'test' && ctx.message?.contains('debug')"
      }
    }
  ]
}
```
{% include copy-curl.html %}

Use the following `_simulate` request to test the pipeline:

```json
POST _ingest/pipeline/advanced_log_pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "status": 503,
        "message": "Server unavailable"
      }
    },
    {
      "_source": {
        "env": "test",
        "message": "debug log output"
      }
    },
    {
      "_source": {
        "status": 200,
        "message": "OK"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> In the response, note that the first document has an `env: production` and `severity: major` fields added. The second document is dropped. The third document has an `env: production` field added:

Reviewer: ```suggestion
In the response, note that the first document has the `env: production` and `severity: major` fields added. The second document is dropped. The third document has an `env: production` field added:
```

```json
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "severity": "major",
          "message": "Server unavailable",
          "env": "production",
          "status": 503
        },
        "_ingest": {
          "timestamp": "2025-04-23T10:51:46.795026554Z"
        }
      }
    },
    null,
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "message": "OK",
          "env": "production",
          "status": 200
        },
        "_ingest": {
          "timestamp": "2025-04-23T10:51:46.795048304Z"
        }
      }
    }
  ]
}
```

## Null-safe notation

> Use null-safe navigation notation (`?.`) to check if the field is `null`. Note that this notation can return `null` silently, therefore we recommend to first check if the returned value is `null` and then use operations like `.contains` or `==`.

Reviewer: ```suggestion
Use null-safe navigation notation (`?.`) to check whether the field is `null`. Note that this notation can return `null` silently; therefore, we recommend first checking whether the returned value is `null` and then using operations like `.contains` or `==`.
```

> Not safe syntax:

Reviewer: ```suggestion
Unsafe syntax:
```

```
"if": "ctx.message?.contains('debug')"
```

> If the `message` field does not exist in the document, this request returns a `null_pointer_exception` with a message `Cannot invoke "Object.getClass()" because "value" is null`.

Reviewer: ```suggestion
If the `message` field does not exist in the document, this request returns a `null_pointer_exception` with the message `Cannot invoke "Object.getClass()" because "value" is null`.
```
